Add Strict Effects to Strict Mode (#4362)
reactjs/reactjs.org: content/docs/strict-mode.md

![](../images/blog/warn-legacy-context-in-strict-mode.png)

Read the [new context API documentation](/docs/context.html) to help migrate to the new version.

### Detecting unsafe effects {#detecting-unsafe-effects}

> Conceptually, React effects include three phases:
> * The **create** phase is called for all effects when the effect is created.
> * The **update** phase is called when the dependencies change, if applicable.
> * The **destroy** phase is called for all effects when the effect is destroyed.

Reviewer: Love this clarification and tying it back to the historical behavior with mounting and unmounting 👍🏻 

Though is there a difference for React developers between update and create?

```suggestion
Conceptually, React effects include three phases:
* The **create** phase is called for all effects when the effect is created i.e. when dependencies change.
* The **destroy** phase is called for all effects when the effect is destroyed.
```
The wording can be polished. It's more about reducing it to two phases: create and destroy.

Author: Removed this section

In the past, React has only called the create phase once when a component mounts:

```
* React mounts the component.
  * Layout effects are created.
  * Effect effects are created.
```

and the destroy phase once when a component unmounts:

```
* React unmounts the component.
  * Layout effects are destroyed.
  * Effect effects are destroyed.
```

> Note:
>
> It's common to hear effects being "mounted" and "unmounted". This is because effects have historically been created when a component was mounted, and destroyed when a component was unmounted. This can cause confusion, so we refer to them as "created" and "destroyed" here instead.

> In the future, we'd like to add features to React which would allow a component to mount without immediately creating effects (such as pre-rendering), or to destroy effects in already-mounted components (such as when a component isn't visible). These features will add better performance and resource management out-of-the-box to React, but require effects to be decoupled from the component lifecycle and resilient to being created and destroyed multiple times in a component.

Reviewer: I went back and forth whether this should mean 
```suggestion
In the future, we'd like to add features to React which would allow a component to mount without immediately creating effects (such as pre-rendering), or to destroy effects in already-mounted components (such as when a component becomes invisible). These features will add better performance and resource management out-of-the-box to React, but require effects to be decoupled from the component lifecycle and resilient to being created and destroyed multiple times in a component.
```
or not.

But I guess it makes sense to mean "is invisible" considering the "pre-rendering" scenario where we would run effects (knowing that something cannot be visible), committing it, checking visibility, noticing that it isn't, and therefore destroying the effect (just guessing here about the semantics of the Offscreen API).
It's just such a new concept (especially considering focus managment and different definitions of what is considered "visible") that I felt like pointing out that this paragraph might be confusing. 

Though you do say "pre-rendering **or** destroy in already-mounted components" so it sounds like completely separate scenarios in which case I'm wondering why we would ever create effects in a component that never was visible.

Author: Would you say something is invisible if it was covered by a modal or dialog? How about "such as when a component is no longer visible on screen"?

Author: Resolving, re-wrote this section so no longer relevant.
